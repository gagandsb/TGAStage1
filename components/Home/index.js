export { default } from './container/Home.container';
