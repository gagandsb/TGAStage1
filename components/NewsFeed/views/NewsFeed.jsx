import React, { Component } from 'react';
import styles from '../styles/NewsFeed.style';
import withStyles from '../../hoc/withStyles';
import TimeAgo from 'javascript-time-ago';
import en from 'javascript-time-ago/locale/en';

class NewsFeed extends Component {
  constructor(props) {
    super(props);
    this.state = {
      upVotes: 0,
    };
  }

  renderListItems = (item, index) => {
    const { title, num_comments, objectID, url, author, created_at } = item;
    TimeAgo.addLocale(en);
    const timeAgo = new TimeAgo('en-US');
    const createdAt = timeAgo.format(new Date(created_at));
    return (
      <div className="feeds-wrapper">
        <div className="row feeds-container" key={index}>
          <div className="desktop-comments">
            <div className="comments-column">
              <span>{num_comments}</span>
            </div>
            <div className="upvote-column">
              <span>{this.getLocalStorageValue(objectID)}</span>
              <img
                src="/static/images/arrow-up.gif"
                alt="up arrow"
                className="arrow"
                onClick={() => this.setLocalStorage(objectID)}
              />
            </div>
          </div>
          <div className="col news-title">
            <span className="news-text">{title}</span>
            <a href={url} className="news-subtext-url" target="_blank">
              ({url})
            </a>
            <span className="news-subtext-author">by {author}</span>
            <span className="news-subtext-time">{createdAt}</span>
          </div>
        </div>
        <div className="row mobile-comments" key={index}>
          <div className="comments-column">
            <span>{num_comments}</span>
          </div>
          <div className="upvote-column">
            <span>{this.getLocalStorageValue(objectID)}</span>
            <img
              src="/static/images/arrow-up.gif"
              alt="up arrow"
              className="arrow"
              onClick={() => this.setLocalStorage(objectID)}
            />
          </div>
        </div>
      </div>
    );
  };

  setLocalStorage = (key) => {
    const items = this.getAllNewsData('NewsUpVotes');
    let results = [];
    let item = {};
    let val = 1;
    let index = -1;
    if (items && items.length) {
      results = items;
      const object = results.find((i) => i.name === key);
      index = results.findIndex((i) => i.name === key);
      val = object ? object.value + 1 : 1;
    }
    item.name = key;
    item.value = val;
    if (index !== -1) {
      results[index] = item;
    } else {
      results.push(item);
    }
    localStorage.setItem('NewsUpVotes', JSON.stringify(results));
    const { upVotes } = this.state;
    this.setState({ upVotes: upVotes + 1 });
  };

  getLocalStorageValue = (key) => {
    const results = this.getAllNewsData();
    if (results && results.length) {
      const item = results.find((i) => i.name === key);
      return item && item.name ? item.value : 0;
    }
    return 0;
  };

  getAllNewsData = () => {
    if (this.isClient()) {
      const items = window.localStorage.getItem('NewsUpVotes');
      return JSON.parse(items);
    }
    return [];
  };

  isClient() {
    return typeof window !== 'undefined';
  }

  render() {
    const { feed, className } = this.props;
    return (
      <div className={className}>
        {feed.map((val, index) => this.renderListItems(val, index))}
      </div>
    );
  }
}

export default withStyles(NewsFeed, styles);
export { NewsFeed as NewsFeedVanilla };
