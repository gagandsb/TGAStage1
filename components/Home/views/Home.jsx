import React, { Component } from 'react';
import fetch from 'isomorphic-unfetch';
import NewsFeed from '../../NewsFeed';
import styles from '../styles/Home.style';
import withStyles from '../../hoc/withStyles';

class Home extends Component {
  constructor(props) {
    super(props);
    this.state = {
      pagesFeed: [],
      page: 0,
    };
  }

  static async getInitialProps() {
    const url = `https://hn.algolia.com/api/v1/search?query=india&tags=story&page=0`;
    const res = await fetch(url);
    const results = await res.json();
    return {
      feed: results.hits,
    };
  }

  fetchNewsFeed = (page = 0) => {
    const url = `https://hn.algolia.com/api/v1/search?query=india&tags=story&page=${page}`;
    fetch(url)
      .then((res) => res.json())
      .then((res) => {
        this.setState({ pagesFeed: res.hits });
      });
  };

  renderNextPage = () => {
    const { page } = this.state;
    const newPage = page + 1;
    this.setState({ page: newPage });
    this.fetchNewsFeed(newPage);
  };

  render() {
    const { pagesFeed } = this.state;
    const { className, feed = [] } = this.props;
    const data = pagesFeed && pagesFeed.length ? pagesFeed : feed;
    return (
      <div className={className}>
        <div className="container-fluid">
          <div className="hacker-news">
            <div className="row header-row">
              <div className="col">
                <a href="#" className="header-text">
                  Hacker News
                </a>
                <a href="#" className="header-subtext">
                  <span>top |</span>
                </a>
                <a href="#" className="header-subtext">
                  <span> new</span>
                </a>
              </div>
            </div>
            <div className="row">
              <NewsFeed feed={data} className="news-wrapper" />
            </div>
            <div className="row footer-row">
              <div className="col">
                <a
                  href="javascript:void(0);"
                  onClick={this.renderNextPage}
                  className="more-link"
                >
                  More
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default withStyles(Home, styles);
export { Home as HomeVanilla };
