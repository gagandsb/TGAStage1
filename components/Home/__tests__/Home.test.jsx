import React from 'react';
import { shallow } from 'enzyme';
import { HomeVanilla as Home } from '../views/Home';

describe('Home component', () => {
  let props;
  let component;
  beforeEach(() => {
    props = {
      className: '',
      feed: [],
    };
    component = shallow(<Home {...props} />);
  });

  it('should render correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
