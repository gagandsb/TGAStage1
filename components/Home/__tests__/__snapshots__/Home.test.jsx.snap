// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component should render correctly 1`] = `
<div
  className=""
>
  <div
    className="container-fluid"
  >
    <div
      className="hacker-news"
    >
      <div
        className="row header-row"
      >
        <div
          className="col"
        >
          <a
            className="header-text"
            href="#"
          >
            Hacker News
          </a>
          <a
            className="header-subtext selected"
            href="#"
            onClick={[Function]}
          >
            <span>
              top
            </span>
          </a>
          <span>
             | 
          </span>
          <a
            className="header-subtext "
            href="#"
            onClick={[Function]}
          >
            <span>
              new
            </span>
          </a>
        </div>
      </div>
      <div
        className="row"
      >
        <Styled(NewsFeed)
          className="news-wrapper"
          feed={Array []}
        />
      </div>
      <div
        className="row footer-row"
      >
        <div
          className="col"
        >
          <a
            className="more-link"
            href="#"
            onClick={[Function]}
          >
            More
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
